chain/index/ddls.go

```diff
@@ -70,13 +70,17 @@ func preparedStatementMapping(ps *preparedStatements) map[**sql.Stmt]string {
 		&ps.removeEthHashesOlderThanStmt:          "DELETE FROM eth_tx_hash WHERE inserted_at < datetime('now', ?)",
 		&ps.updateTipsetsToRevertedFromHeightStmt: "UPDATE tipset_message SET reverted = 1 WHERE height >= ?",
 		&ps.updateEventsToRevertedFromHeightStmt:  "UPDATE event SET reverted = 1 WHERE message_id IN (SELECT message_id FROM tipset_message WHERE height >= ?)",
-		&ps.isTipsetMessageNonEmptyStmt:           "SELECT EXISTS(SELECT 1 FROM tipset_message LIMIT 1)",
+		&ps.isIndexEmptyStmt:                      "SELECT NOT EXISTS(SELECT 1 FROM tipset_message LIMIT 1)",
 		&ps.getMinNonRevertedHeightStmt:           "SELECT MIN(height) FROM tipset_message WHERE reverted = 0",
 		&ps.hasNonRevertedTipsetStmt:              "SELECT EXISTS(SELECT 1 FROM tipset_message WHERE tipset_key_cid = ? AND reverted = 0)",
 		&ps.updateEventsToRevertedStmt:            "UPDATE event SET reverted = 1 WHERE message_id IN (SELECT message_id FROM tipset_message WHERE tipset_key_cid = ?)",
 		&ps.updateEventsToNonRevertedStmt:         "UPDATE event SET reverted = 0 WHERE message_id IN (SELECT message_id FROM tipset_message WHERE tipset_key_cid = ?)",
 		&ps.getMsgIdForMsgCidAndTipsetStmt:        "SELECT message_id FROM tipset_message WHERE tipset_key_cid = ? AND message_cid = ? AND reverted = 0",
-		&ps.insertEventStmt:                       "INSERT INTO event (message_id, event_index, emitter_addr, reverted) VALUES (?, ?, ?, ?)",
+		&ps.insertEventStmt:                       "INSERT INTO event (message_id, event_index, emitter_addr, reverted) VALUES (?, ?, ?, ?) ON CONFLICT (message_id, event_index) DO UPDATE SET reverted = 0",
```

Comment:
some kind of test that exercises this and demonstrates it working well would be nice, also does the query plan show it behaving nicely? I can imagine this getting a bit weird matching for conflicts

Reply:
@rvagg 

- Will be writing a test for this.

- While sqliite does not generate query plans for `INSERT STATEMENTS`, the query plan for an equivalent query says this will be performant as we have a covering index for `(message_id, event_index)`

```
EXPLAIN QUERY PLAN
UPDATE event SET reverted = 0
WHERE message_id = 1 AND event_index = 2;

EXPLAIN QUERY PLAN
INSERT INTO event (message_id, event_index, emitter_addr, reverted) 
SELECT 1, 2, 'a', 0
WHERE NOT EXISTS (
    SELECT 1 FROM event 
    WHERE message_id = 1 AND event_index = 2
);
```

Reply:
This is because we have an `UNIQUE` constraint on `(message_id, event_index)`

Reply:
Done.